src/protocol/mod.rs

```diff
@@ -0,0 +1,31 @@
+mod bpp;
+mod message;
+
+use crate::errors::ProtocolError;
+
+use std::io::Read;
+
+pub trait Encode {
+    // Consume the entity and return it's BE byte representation.
+    fn encode_to_bytes(self) -> Result<Vec<u8>, ProtocolError>;
+
+    // Calculate the size of BE byte representation.
+    fn encoded_len(&self) -> usize;
```

Comment:
idea: if methods are in Encode trait, should they start with encoded_* ? If we ensure that it isn't possible to have 2 different methods in different traits with the same name, it could look better imo

Reply:
I didn't want to use size() and to_bytes() because those are provided by some builtin traits

Reply:
But I agree we can rethink those names

Reply:
After offline discussion we will use more standard names. 
https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#fully-qualified-syntax-for-disambiguation-calling-methods-with-the-same-name